signac/core/collection_api.py

```diff
@@ -0,0 +1,491 @@
+
+    def pop(self, key, default=None):
+        self.load()
+        with self._safe_sync():
+            ret = self._data.pop(key, default)
+            self.sync()
+        return ret
+
+    def popitem(self, key, default=None):
+        self.load()
+        with self._safe_sync():
+            ret = self._data.pop(key, default)
+            self.sync()
+        return ret
+
+    def clear(self):
+        self.load()
+        with self._safe_sync():
+            self._data = {}
+            self.sync()
+
+    def update(self, mapping):
+        self.load()
+        with self._safe_sync():
+            with self._suspend_sync():
+                for key, value in mapping.items():
+                    self[key] = self.from_base(data=value, parent=self)
+            self.sync()
+
+    def setdefault(self, key, default=None):
+        self.load()
+        with self._safe_sync():
+            with self._suspend_sync():
+                ret = self._data.setdefault(key, self.from_base(data=default, parent=self))
+            self.sync()
+        return ret
+
+
+class SyncedAttrDict(_SyncedDict):
+
+    def __getattr__(self, name):
+        try:
+            return super().__getattribute__(name)
+        except AttributeError:
+            if name.startswith('__'):
+                raise
+            try:
+                return self.__getitem__(name)
+            except KeyError as e:
+                raise AttributeError(e)
+
+    def __setattr__(self, key, value):
+        try:
+            super().__getattribute__('_data')
+        except AttributeError:
+            super().__setattr__(key, value)
+        else:
+            if key.startswith('__') or key in self._PROTECTED_KEYS:
+                super().__setattr__(key, value)
+            else:
+                self.__setitem__(key, value)
+
+    def __delattr__(self, key):
+        if key.startswith('__') or key in self._PROTECTED_KEYS:
+            super().__delattr__(key)
+        else:
+            self.__delitem__(key)
+
+
+class SyncedList(SyncedCollection, MutableSequence):
+
+    def __init__(self, **kwargs):
+        data = kwargs.pop('data', None)
+        super().__init__(**kwargs)
+        if data is None:
+            self._data = []
+        else:
+            self._data = [self.from_base(data=value, parent=self) for value in data]
+
+    def to_base(self):
+        converted = list()
+        for value in self._data:
+            if isinstance(value, SyncedCollection):
+                converted.append(value.to_base())
+            else:
+                converted.append(value)
+        return converted
+
+    def reset(self, data=None):
+        if data is None:
+            data = []
+        if isinstance(data, Sequence) and not isinstance(data, str):
+            backup = copy(self._data)
+            try:
+                with self._suspend_sync():
+                    for i in range(min(len(self), len(data))):
+                        if data[i] == self._data[i]:
```

Comment:
What is the purpose of checking equality and resetting/constructing items individually based on whether their index is before or after `len(self._data)`? I would have expected this block would just delete `self._data` and build a new set of synced items from scratch.

Reply:
I agree that this seems unnecessary and made similar comments here and above. I think there's an aspect to this that's optimization-related, which is whether we can avoid rebuilding parts of a list if they're going to be identical before and after the reset. However, this solution (particularly for lists) is probably suboptimal anyway since there are ordering constraints.

Reply:
It also works something like `dfs_update`. So every time data is loaded it does not create a different instance. But I agree with @vyasr  as it takes order into consideration.

Reply:
I have suggested a comment in the code that marks this as a place for potential optimization in the future. I will mark this as resolved for now.